0634-cpufreq-Add-Hardware-P-State-HWP-driver.patch

```diff
@@ -0,0 +1,761 @@
++static bool feature_hwp_energy_perf;
++static bool feature_hwp_pkg_level_ctl;
++static bool feature_hwp_peci;
++
++static bool feature_hdc;
++
++__initdata bool opt_cpufreq_hwp = false;
++__initdata bool opt_cpufreq_hdc = true;
++
++#define HWP_ENERGY_PERF_BALANCE         0x80
++#define IA32_ENERGY_BIAS_BALANCE        0x7
++#define IA32_ENERGY_BIAS_MAX_POWERSAVE  0xf
++#define IA32_ENERGY_BIAS_MASK           0xf
++
++union hwp_request
++{
++    struct
++    {
++        uint64_t min_perf:8;
++        uint64_t max_perf:8;
++        uint64_t desired:8;
++        uint64_t energy_perf:8;
++        uint64_t activity_window:10;
++        uint64_t package_control:1;
++        uint64_t reserved:16;
++        uint64_t activity_window_valid:1;
++        uint64_t energy_perf_valid:1;
++        uint64_t desired_valid:1;
++        uint64_t max_perf_valid:1;
++        uint64_t min_perf_valid:1;
++    };
++    uint64_t raw;
++};
++
++struct hwp_drv_data
++{
++    union
++    {
++        uint64_t hwp_caps;
++        struct
++        {
++            uint64_t highest:8;
++            uint64_t guaranteed:8;
++            uint64_t most_efficient:8;
++            uint64_t lowest:8;
++            uint64_t reserved:32;
++        } hw;
++    };
++    union hwp_request curr_req;
++    uint16_t activity_window;
++    uint8_t minimum;
++    uint8_t maximum;
++    uint8_t desired;
++    uint8_t energy_perf;
++};
++DEFINE_PER_CPU_READ_MOSTLY(struct hwp_drv_data *, hwp_drv_data);
++
++#define hwp_err(...)     printk(XENLOG_ERR __VA_ARGS__)
++#define hwp_info(...)    printk(XENLOG_INFO __VA_ARGS__)
++#define hwp_verbose(...)                   \
++({                                         \
++    if ( cpufreq_verbose )                 \
++        printk(XENLOG_DEBUG __VA_ARGS__);  \
++})
++
++static int cf_check hwp_governor(struct cpufreq_policy *policy,
++                                 unsigned int event)
++{
++    int ret;
++
++    if ( policy == NULL )
++        return -EINVAL;
++
++    switch ( event )
++    {
++    case CPUFREQ_GOV_START:
++    case CPUFREQ_GOV_LIMITS:
++        ret = 0;
++        break;
++
++    case CPUFREQ_GOV_STOP:
++    default:
++        ret = -EINVAL;
++        break;
++    }
++
++    return ret;
++}
++
++static struct cpufreq_governor hwp_cpufreq_governor =
++{
++    .name          = XEN_HWP_GOVERNOR,
++    .governor      = hwp_governor,
++};
++
++static int __init cf_check cpufreq_gov_hwp_init(void)
++{
++    return cpufreq_register_governor(&hwp_cpufreq_governor);
++}
++__initcall(cpufreq_gov_hwp_init);
++
++bool __init hwp_available(void)
++{
++    unsigned int eax, ecx, unused;
++    bool use_hwp;
++
++    if ( boot_cpu_data.cpuid_level < CPUID_PM_LEAF )
++    {
++        hwp_verbose("cpuid_level (%u) lacks HWP support\n",
++                    boot_cpu_data.cpuid_level);
++        return false;
++    }
++
++    if ( boot_cpu_data.cpuid_level < 0x16 )
++    {
++        hwp_info("HWP disabled: cpuid_level %x < 0x16 lacks CPU freq info\n",
```

Comment:
@jandryuk `0x%x` ?

Reply:
Yes, good catch.  Thanks.  I'll also change the cpuid_level printing a few lines up from `%u` to `%#x` for consistency.